src/core/enum_type.hpp

```diff
@@ -75,4 +75,56 @@ debug_inline constexpr void ToggleFlag(T &x, const T y)
 	}
 }
 
+template <typename Tenum, typename Tstorage>
+class EnumBitSet {
+public:
+	constexpr EnumBitSet() : data(0) {}
+	constexpr EnumBitSet(Tenum other) : data(to_underlying(other)) {}
+	constexpr EnumBitSet(Tstorage data) : data(data) {}
+
+	constexpr EnumBitSet(const EnumBitSet &) = default;
+
+	constexpr auto operator <=>(const EnumBitSet &) const noexcept = default;
+
+	inline constexpr EnumBitSet &set(Tenum value)
+	{
+		this->data |= (1U << to_underlying(value));
+		return *this;
+	}
+
+	inline constexpr EnumBitSet &clear(Tenum value)
+	{
+		this->data &= ~(1U << to_underlying(value));
+		return *this;
+	}
+
+	inline constexpr EnumBitSet &toggle(Tenum value)
+	{
+		if (this->test(value)) {
+			return this->clear(value);
+		} else {
+			return this->set(value);
+		}
+	}
+
+	inline constexpr bool test(Tenum value) const
+	{
+		return (this->data & (1U << to_underlying(value))) != 0;
+	}
+
+	inline constexpr EnumBitSet operator |(Tenum value) const
+	{
+		EnumBitSet other{*this};
+		return other.set(value);
+	}
+
+	inline constexpr Tstorage base() const noexcept
```

Comment:
I reckon the endian writer couldn't be friended, right?

Reply:
Not sure, but having an explicit method to get to the underlying storage is similar to what we have with StrongType.

Reply:
True. It was that the only reason I could think of you would need access to the base would be such serialisation code. Anything else shouldn't need access to it, whereas that's quite different for the strong types.

I'm fine with it the way it is right now. It's just a "might be worth it, if it's simple" to prevent misuse.

Reply:
Also I went a bit further to see where this leads ("I have a patch for that..."), and .base() is needed elsewhere, e.g. for values that get reported to NewGRF or to SetDParam()